### PR TITLE
HWKAPM-560 Update instance count when criteria is changed, including …

### PR DESCRIPTION
**Pull request by objectiser**
…the URI and optional operation related to the selected node

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay Would you be able to review this PR at some point - not urgent, but will need to build the PR and use the booker example test data you have used previously. Just need to check that when the Distributed Tracing diagram root node is selected, that the instance count is correct - i.e. should be same as instance count on the first node in the diagram, and also should relate to the number of lines in the table when the instance button is pressed.

**Comment by objectiser** [comment]
@ammendonca Once back from PTO, would you mind just reviewing my UI changes to see whether they are ok, or if there is a better way to achieve the same results? Thanks.

File: server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java [hunk 1]
```diff
@@ -454,6 +454,42 @@ public void getTraceCompletionCount(
...
+
+            long count = analyticsService.getTraceCompletionCount(
+                    securityProvider.validate(tenantId, context.getUserPrincipal().getName()), criteria);
+
+            log.tracef("Got trace completion count: criteria [%s] = [%s]", criteria, count);
+
+            response.resume(Response.status(Response.Status.OK).entity(count).type(APPLICATION_JSON_TYPE)
+                    .build());
+
+        } catch (Throwable e) {
```
Comment (pavolloffay) [inline server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java]: Isn't there a more specific exception?

File: server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java [hunk 2]
```diff
@@ -524,6 +560,42 @@ public void getTraceCompletionFaultCount(
...
+    @POST
+    @Path("trace/completion/faultcount")
+    @Produces(APPLICATION_JSON)
+    @ApiOperation(
+            value = "Get the number of trace instances that returned a fault",
+            response = Long.class)
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "Success"),
+            @ApiResponse(code = 500, message = "Internal server error") })
+    public void getTraceCompletionFaultCount(
```
Comment (pavolloffay) [inline server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java]: There is a GET and a POST handler for `trace/completion/faultcount`. These two could reuse some code.

Comment (objectiser) [inline server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java]: Yes, this is a more general change that needs to be made to all GET/POST alternatives. So if ok will raise a separate jira to cover all.

Comment (pavolloffay) [inline server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java]: A separate jira is fine :+1:

File: server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java (hunk 1 above)
Comment (objectiser) [inline server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java]: Ah caught again :) - not really, I've just followed the same style as other REST services in hawkular (e.g. https://github.com/hawkular/hawkular-alerts/blob/master/hawkular-alerts-rest/src/main/java/org/hawkular/alerts/rest/ActionsHandler.java#L203, https://github.com/hawkular/hawkular-alerts/blob/master/hawkular-alerts-rest/src/main/java/org/hawkular/alerts/rest/EventsHandler.java#L120) although looks like inventory and metrics now use a more async style. So its a general issue on the REST services - so better dealt with as a separate issue aswell?

Comment (pavolloffay) [inline server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java]: I think that here is better to catch checked exceptions and for unchecked use exception mappers.

File: server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java (hunk 2 above)
Comment (objectiser) [inline server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java]: https://issues.jboss.org/browse/HWKAPM-569

File: server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java (hunk 1 above)
Comment (objectiser) [inline server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java]: https://issues.jboss.org/browse/HWKAPM-570

**Comment by objectiser** [comment]
@pavolloffay Thanks Pavol - have created two separate jiras for those issues, as they are general issues with REST handlers.

**Comment by pavolloffay** [comment]
Thanks :+1:

**Pull request merged**